File: src/main/java/io/veridia/jsonlogic/OperatorRegistry.java
package io.veridia.jsonlogic;

import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

public class OperatorRegistry {
    private final Map<String, Operator> ops = new ConcurrentHashMap<>();

    public void register(Operator impl) {
        ops.put(impl.key(), impl);
    }

    public Operator get(String name) {
        Operator op = ops.get(name);
        if (op == null) {
            throw new UnsupportedOperationException("Operator not implemented: " + name);
        }
        return op;
    }
}

File: src/main/scala/io/veridia/jsonlogic/OperatorRegistry.scala
package io.veridia.jsonlogic

import scala.collection.concurrent.TrieMap

/** Lock-free, thread-safe operator dispatch table — every `apply`/`check` call may run
  * concurrently across tenants sharing one `JsonLogic` instance.
  */
final class OperatorRegistry:
  private val ops = TrieMap.empty[String, Operator]

  def register(impl: Operator): Unit = ops.put(impl.key(), impl)

  def get(name: String): Operator =
    ops.getOrElse(name, throw new UnsupportedOperationException(s"Operator not implemented: $name"))
